thunder/transforms/materialization.py

```diff
@@ -13,6 +13,83 @@
     from thunder.core.module import ThunderModule
 
 
+def module_init_from_original_module_init(
+    transform: MaterializationTransform, tm: ThunderModule, *, non_persistent_only: bool = False
+):
+    if non_persistent_only:
+        non_persistent_buffers_set = set()
+        for name, submodule in tm._model.named_modules():
+            prefix = f"{name}." if name else name
+            non_persistent_buffers_set.update(f"{prefix}{bname}" for bname in submodule._non_persistent_buffers_set)
+
+        def neg_filter(n):
+            return n not in non_persistent_buffers_set
+
+    else:
+
+        def neg_filter(n):
+            return False
+
+    shared_names = tm._get_shared_names()
+    processed_names = set()
+
+    # Shared parameters in PyTorch eager are parameters of module which have different name but share the underlying tensor.
+    # For shared parameter, we replace all occurence shared parameter with its corresponding `base` parameter.
+    # In our implementation `base` parameter is the parameter and corresponding name which we see the first time while
+    # iterating our parameters (see below). We track subsequent parameter which share the underlying Tensor with this `base` parameter
+    # in `shared_params_name` dictionary.
+
+    for module_name, _ in tm._model.named_modules():
+        prefix = module_name if not module_name else f"{module_name}."
+        submodule = tm.get_submodule(module_name)
+
+        # Materialize meta-parameters on-device if necessary.
+        # This is done before sharding in case the materialization logic depends on the tensor shape.
+        # The tradeoff is that all of a module's direct parameters need to fit in device.
+        # Each module only initializes its own parameters and not those of its children (recurse=False)
+        if any(
+            chain(
+                (
+                    (n in transform.have_materialized) and not neg_filter(n)
+                    for n, _ in submodule.named_parameters(recurse=False, prefix=module_name)
+                ),
+                (
+                    (n in transform.have_materialized) and not neg_filter(n)
+                    for n, _ in submodule.named_buffers(recurse=False, prefix=module_name)
+                ),
+            )
+        ):
+            # we use a copy to let the user's module alone
+            module_copy = copy.copy(submodule)
+            module_copy._parameters = module_copy._parameters.copy()
+            module_copy._buffers = module_copy._buffers.copy()
+            module_copy._modules = module_copy._modules.__class__()
+
+            # we need to initialize the module unless all parameters are duplicatess
+            need_init = not all(
+                bool(shared_names[n] & processed_names) | neg_filter(n)
```

Comment:
could you make this line more Python-like? e.g. it seems to me that we can use `and` and `or` keywords instead of `&` and `|` respectively

Reply:
Fair for the `|`, it should be `or`. The `&` however is a set operation. Thanks for pointing it out, having the `or` there is much better!